src/v/ssx/semaphore.h

```diff
@@ -59,9 +60,9 @@ class mutex {
     using duration = typename ss::semaphore::duration;
     using time_point = typename ss::semaphore::time_point;
 
-    // TODO constructor to pass through name & change callers.
-    mutex()
-      : _sem(1, "mutex") {}
+    // All semaphores and mutexes are named to aid in debugging.
+    explicit mutex(ss::sstring name)
```

Comment:
pass by string_view?

Reply:
an alternative approach here to avoid the extra `ntp.to_name(prefix)` interface would be to create a constructor here that takes a format strings and arguments. then you could write:

`mutex("{}-{}", "mutex_name", ntp);`

Reply:
Yeah, this was the approach I disliked the least. The idea of objects knowing how to do to_string() in general makes sense. This is an additional convenience that allows you to pass in a prefix. To me this is ok; no smells, no layering violation.

Reply:
afaik to_string is really intended for numerics. in any case, ntp already converts to string: `fmt::format("{}", ntp)`

Reply:
:+1:  emphasis on the _idea of .. in general_ (not language specific). Was just saying the choice I made here feels congruent to me (IOW: how to summarize a Foo compactly has more to do with the Foo, than it does with mutex, keep the latter's interface lean). Anti-argument is something like "needing a short name for a Foo is specific to mutex, not Foo". Potato potato.  :shrug:   If you prefer the constructor LMK.

Reply:
Returning a string guarantees inefficiency by requiring allocation. IMHO specialising `fmt::formatter` is the best implementation, although I see no reason not to rely on `operator<<` and a free function for adding a prefix.

Reply:
> Thanks. I'll switch to `fmt::formatter` in next update. For efficiency, I was leaning on short string optimization (fewer allocs at least) and the fact that these are not generally fast paths (creating semaphores).

I guess an ntp often falls outside of 15 bytes. It's probably not a big deal in this case, I was just discussing interface design in general. Death by a thousand cuts and all that.

Reply:
Yeah for the fmt::formatter approach, it ends up being a little verbose since that doesn't include plumbing for a prefix.. so more code overall. I'll play around with it for a bit.

Reply:
> If you prefer the constructor LMK.

i don't have a strong opinion, other than it's useful to avoid a custom string formatting function off the widely used ntp just for naming mutexes.